### ISSUE TITLE
[v2.3x]

### ISSUE THREAD

**Issue by b005t3r** [BowlerHatLLC/feathers#1469]
Word wrap is not handled correctly if a label is inside a list item renderer:

![word_wrap_problem](https://cloud.githubusercontent.com/assets/757150/18947988/95ff9388-8635-11e6-9794-cdfbc82d1960.jpg)

As you can see, line breaks are handled correctly (first element), but if a word wrap is needed, renders are not resized to the right size.

Here's the renderer code:
```as3
class OptionItemRenderer extends LayoutGroup implements IListItemRenderer {
    public static const STANDARD_POLYGON_DATA:String            = "463.655,482.267 413.939,482.267 408.675,471.719 121.545,471.719 116.582,482.267 54.939,482.267 52.109,471.719 33.188,471.719 29.582,482.267 14.326,482.267 9.749,473.104 9.745,27.962 14.338,18.765 29.333,18.765 34.816,9.750 98.050,9.750 103.187,18.765 422.388,18.765 426.477,9.729 446.421,9.729 449.762,18.765 463.667,18.765 468.261,27.962 468.231,473.104";
    public static const FIRST_HORIZONTAL_REGION_SIZE:Number     = 140;
    public static const SECOND_HORIZONTAL_REGION_SIZE:Number    = 238;
    public static const FIRST_VERTICAL_REGION_SIZE:Number       = 42;
    public static const SECOND_VERTICAL_REGION_SIZE:Number      = 406;

    private static const parsedData:Polygon2D                   = PolygonUtil.parse(STANDARD_POLYGON_DATA);

    public static function getScaledPolygon(newWidth:Number, newHeight:Number):Polygon2D {
        return PolygonUtil.scale9Both(
            parsedData,
            newWidth,
            newHeight,
            FIRST_HORIZONTAL_REGION_SIZE,
            SECOND_HORIZONTAL_REGION_SIZE,
            FIRST_VERTICAL_REGION_SIZE,
            SECOND_VERTICAL_REGION_SIZE
        );
    }

    private var _assetManager:AssetManager;
    private var _localeManager:LocaleManager;

    private var _data:OptionSettings;
    private var _index:int;
    private var _owner:List;

    private var _background:GlowPanel;
    private var _nameLabel:Label;
    private var _descLabel:Label;
    private var _button:GlowFrameTextButton;

    public function OptionItemRenderer(assetManager:AssetManager, localeManager:LocaleManager) {
        _assetManager = assetManager;
        _localeManager = localeManager;

        _nameLabel = new Label();
        _nameLabel.textRendererFactory = function():ITextRenderer {
            var renderer:BitmapFontTextRenderer = new BitmapFontTextRenderer();
            renderer.textFormat = new BitmapFontTextFormat("hero_glow_30pt", NaN, Color.WHITE, TextFormatAlign.LEFT);
            return renderer;
        };
        
        _descLabel = new Label();
        _descLabel.textRendererProperties.wordWrap = true;
        _descLabel.textRendererFactory = function():ITextRenderer {
            var renderer:BitmapFontTextRenderer = new BitmapFontTextRenderer();
            renderer.textFormat = new BitmapFontTextFormat("hero_glow_18pt", NaN, Color.WHITE, TextFormatAlign.JUSTIFY, -6);
            renderer.alpha = 0.95;
            return renderer;
        };

        _button = new GlowFrameTextButton("", _assetManager);

        layout = new AnchorLayout();

        var topOffset:Number    = 30;
        var bottomOffset:Number = 30;
        var edgeOffset:Number   = 25;

        _nameLabel.layoutData   = new AnchorLayoutData(topOffset, edgeOffset + 60, NaN, edgeOffset);
        _descLabel.layoutData   = new AnchorLayoutData(topOffset + 50, edgeOffset + 60, bottomOffset, edgeOffset);
        _button.layoutData      = new AnchorLayoutData(topOffset, edgeOffset);

        addChild(_nameLabel);
        addChild(_descLabel);
        addChild(_button);
    }

    public function get button():GlowFrameTextButton { return _button; }

    public function get data():Object { return _data; }
    public function set data(value:Object):void {
        _data = OptionSettings(value);

        if(_data == null) {
            _nameLabel.text         = "";
            _descLabel.text         = "";
            _button.textField.text  = "";
        }
        else {
            _nameLabel.text         = _data.name;
            _descLabel.text         = _data.description;
            _button.textField.text  = _data.availableValues[_data.currentValueIndex];
        }
    }

    public function get index():int { return _index; }
    public function set index(value:int):void { _index = value; }

    public function get owner():List { return _owner; }
    public function set owner(value:List):void { _owner = value; }

    public function get factoryID():String { return null; }
    public function set factoryID(value:String):void { }

    public function get isSelected():Boolean { return false; }
    public function set isSelected(value:Boolean):void { }

    override public function render(support:RenderSupport, parentAlpha:Number):void {
        redrawBackground();

        support.pushMatrix();
        support.transformMatrix(_background);
        support.blendMode = _background.blendMode;

        _background.render(support, parentAlpha);

        support.blendMode = blendMode;
        support.popMatrix();

        super.render(support, parentAlpha);
    }

    private function redrawBackground():void {
        var w:Number = width;
        var h:Number = height;

        var xOffset:Number = 4.33, xMargin:Number = 4, yOffset:Number = 6;

        if(w == 0 || h == 0)
            return;

        if(_background != null && Math.abs(_background.width - w + 2 * xMargin) < 1 && Math.abs(_background.height - h) < 1)
            return;

        if(_background != null)
            _background.dispose();

        _background = new GlowPanel(
            new Scale3Textures(_assetManager.getTexture("ui/thick_window_edge"), 6, 2),
            new Point2D(xOffset, yOffset),
            _assetManager.getTexture("ui/window_gradient"),
            _assetManager.getTexture("ui/window_pattern"),
            getScaledPolygon(w - xOffset - 2 * xMargin, h - yOffset)
        );

        _background.x = xMargin + xOffset / 2;
        _background.y = yOffset / 2;
    }
}
```

and the list is initialized like so:
```as3
        var listLayout:VerticalLayout           = new VerticalLayout();
        listLayout.hasVariableItemDimensions    = true;
        listLayout.useVirtualLayout             = true;
        listLayout.horizontalAlign              = VerticalLayout.HORIZONTAL_ALIGN_JUSTIFY;
        listLayout.gap                          = LIST_ITEM_SPACING;
        listLayout.lastGap                      = LIST_ITEM_SPACING;

        _list = new List();
        _list.layout = listLayout;
        _list.itemRendererFactory = function():IListItemRenderer {
            var renderer:OptionItemRenderer = new OptionItemRenderer(_assetManager, _localeManager);
            renderer.addEventListener(Event.TRIGGERED, function(event:Event):void {
                dispatchEventWith(OPTION_TRIGGERED_EVENT, false, renderer.data);
            });

            return renderer;
        };

        addChild(_list);
```

**Comment by joshtynjala** [comment]
Just in case I need to run this code again in the future, here it is without the extra dependencies that aren't in vanilla Starling and Feathers:

Create the List:

	this.autoSizeMode = LayoutGroup.AUTO_SIZE_MODE_STAGE;
	this.layout = new AnchorLayout();

	var listLayout:VerticalLayout           = new VerticalLayout();
	listLayout.hasVariableItemDimensions    = true;
	listLayout.useVirtualLayout             = true;
	listLayout.horizontalAlign              = VerticalLayout.HORIZONTAL_ALIGN_JUSTIFY;
	listLayout.gap                          = 10;
	listLayout.lastGap                      = 10;

	var _list:List = new List();
	_list.layoutData = new AnchorLayoutData(0, 0, 0, 0);
	_list.layout = listLayout;
	_list.itemRendererFactory = function():IListItemRenderer {
		var renderer:OptionItemRenderer = new OptionItemRenderer();
		renderer.addEventListener(Event.TRIGGERED, function(event:Event):void {
			//dispatchEventWith(OPTION_TRIGGERED_EVENT, false, renderer.data);
		});

		return renderer;
	};
	
	_list.dataProvider = new ListCollection(
	[
		{
			name: "Game Timer",
			description: "Lorem ipsum dolor sit amet, consectetur\nadipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
			buttonText: "1 day"
		},
		{
			name: "Combat resolution",
			description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
			buttonText: "Manual"
		},
		{
			name: "Plasma Missiles",
			description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
			buttonText: "Enabled"
		},
	])

	addChild(_list);

The item renderer:


	class OptionItemRenderer extends LayoutGroup implements IListItemRenderer {

		private var _data:Object;
		private var _index:int;
		private var _owner:List;

		private var _nameLabel:Label;
		private var _descLabel:Label;
		private var _button:Button;

		public function OptionItemRenderer() {
			
			backgroundSkin = new Quad(1, 1, 0x666666)
			
			_nameLabel = new Label();
			/*_nameLabel.textRendererFactory = function():ITextRenderer {
				var renderer:BitmapFontTextRenderer = new BitmapFontTextRenderer();
				renderer.textFormat = new BitmapFontTextFormat(BitmapFont.MINI, NaN, Color.WHITE, TextFormatAlign.LEFT);
				return renderer;
			};*/

			_descLabel = new Label();
			_descLabel.textRendererProperties.wordWrap = true;
			/*_descLabel.textRendererFactory = function():ITextRenderer {
				var renderer:BitmapFontTextRenderer = new BitmapFontTextRenderer();
				renderer.textFormat = new BitmapFontTextFormat(BitmapFont.MINI, NaN, Color.WHITE, TextFormatAlign.JUSTIFY, -6);
				renderer.alpha = 0.95;
				return renderer;
			};*/

			_button = new Button();

			layout = new AnchorLayout();

			var topOffset:Number    = 30;
			var bottomOffset:Number = 30;
			var edgeOffset:Number   = 25;

			_nameLabel.layoutData   = new AnchorLayoutData(topOffset, edgeOffset + 60, NaN, edgeOffset);
			_descLabel.layoutData   = new AnchorLayoutData(topOffset + 50, edgeOffset + 60, bottomOffset, edgeOffset);
			_button.layoutData      = new AnchorLayoutData(topOffset, edgeOffset);

			addChild(_nameLabel);
			addChild(_descLabel);
			addChild(_button);
		}

		public function get button():Button { return _button; }

		public function get data():Object { return _data; }
		public function set data(value:Object):void {
			_data = value

			if(_data == null) {
				_nameLabel.text         = "";
				_descLabel.text         = "";
				_button.label  = "";
			}
			else {
				_nameLabel.text         = _data.name;
				_descLabel.text         = _data.description;
				_button.label  			= _data.buttonText;
			}
		}

		public function get index():int { return _index; }
		public function set index(value:int):void { _index = value; }

		public function get owner():List { return _owner; }
		public function set owner(value:List):void { _owner = value; }

		public function get factoryID():String { return null; }
		public function set factoryID(value:String):void { }

		public function get isSelected():Boolean { return false; }
		public function set isSelected(value:Boolean):void { }
	}

**Comment by joshtynjala** [comment]
I have confirmed that your code works correctly in Feathers 3, but not in Feathers 2.3. This was a bug in AnchorLayout where using percentHeight with word-wrapped text would result in an incorrect height measurement. As I recall, AnchorLayout wasn't setting a proper maxWidth, so it was wrapping at a larger width (or not wrapping at all).

**Comment by b005t3r** [comment]
So is it fixed in 2.3x? I can't switch to Starling 2.0, so switching to Feathers 3.0 is also not an option.

**Comment by joshtynjala** [comment]
No, this bug is not fixed in Feathers 2.3.

**Comment by b005t3r** [comment]
And can you fix it? :) I use Feathers 2.3x for a commercial product which is currently live and selling, so I can't just, well, accept there's a bug :)

Is there a workaround, anything I can do about it?

**Comment by joshtynjala** [comment]
I no longer maintain the 2.3.x branch, but if you're interested, I'm happy to offer custom maintenance on a legacy version of Feathers as a premium service.

As a workaround, you can probably manually calculate the width of the Label with wordWrap, instead of using the left and right properties on AnchorLayoutData. Assuming that the width of the item renderer is set explicitly, then the item renderer's actualWidth variable will provide the width in an override of draw(). If you calculate the Label's width before calling super.draw(), it should be enough.

**Comment by b005t3r** [comment]
OK, I'll try the actualWidth workaround first.

**Comment by b005t3r** [comment]
OK, I modified the renderer like so and now it works fine (I'm posting that just in case somebody runs into a similar issue):

```as3
class OptionItemRenderer extends LayoutGroup implements IListItemRenderer {
    public static const STANDARD_POLYGON_DATA:String            = "463.655,482.267 413.939,482.267 408.675,471.719 121.545,471.719 116.582,482.267 54.939,482.267 52.109,471.719 33.188,471.719 29.582,482.267 14.326,482.267 9.749,473.104 9.745,27.962 14.338,18.765 29.333,18.765 34.816,9.750 98.050,9.750 103.187,18.765 422.388,18.765 426.477,9.729 446.421,9.729 449.762,18.765 463.667,18.765 468.261,27.962 468.231,473.104";
    public static const FIRST_HORIZONTAL_REGION_SIZE:Number     = 140;
    public static const SECOND_HORIZONTAL_REGION_SIZE:Number    = 238;
    public static const FIRST_VERTICAL_REGION_SIZE:Number       = 42;
    public static const SECOND_VERTICAL_REGION_SIZE:Number      = 406;

    public static const TOP_OFFSET:Number                       = 30;
    public static const BOTTOM_OFFSET:Number                    = 30;
    public static const EDGE_OFFSET:Number                      = 25;

    private static const parsedData:Polygon2D                   = PolygonUtil.parse(STANDARD_POLYGON_DATA);

    public static function getScaledPolygon(newWidth:Number, newHeight:Number):Polygon2D {
        return PolygonUtil.scale9Both(
            parsedData,
            newWidth,
            newHeight,
            FIRST_HORIZONTAL_REGION_SIZE,
            SECOND_HORIZONTAL_REGION_SIZE,
            FIRST_VERTICAL_REGION_SIZE,
            SECOND_VERTICAL_REGION_SIZE
        );
    }

    private var _assetManager:AssetManager;
    private var _localeManager:LocaleManager;

    private var _data:OptionSettings;
    private var _index:int;
    private var _owner:List;

    private var _background:GlowPanel;
    private var _nameLabel:Label;
    private var _descLabel:Label;
    private var _button:GlowFrameTextButton;

    public function OptionItemRenderer(assetManager:AssetManager, localeManager:LocaleManager) {
        _assetManager = assetManager;
        _localeManager = localeManager;

        _nameLabel = new Label();
        _nameLabel.textRendererFactory = function():ITextRenderer {
            var renderer:BitmapFontTextRenderer = new BitmapFontTextRenderer();
            renderer.textFormat = new BitmapFontTextFormat("hero_glow_30pt", NaN, Color.WHITE, TextFormatAlign.LEFT);
            return renderer;
        };
        
        _descLabel = new Label();
        _descLabel.textRendererProperties.wordWrap = true;
        _descLabel.textRendererFactory = function():ITextRenderer {
            var renderer:BitmapFontTextRenderer = new BitmapFontTextRenderer();
            renderer.textFormat = new BitmapFontTextFormat("hero_glow_18pt", NaN, Color.WHITE, TextFormatAlign.JUSTIFY, -6);
            renderer.alpha = 0.95;
            return renderer;
        };

        _button = new GlowFrameTextButton("", _assetManager);

        layout = new AnchorLayout();

        _nameLabel.layoutData   = new AnchorLayoutData(TOP_OFFSET + 6, EDGE_OFFSET + 60, NaN, EDGE_OFFSET);
        _descLabel.layoutData   = new AnchorLayoutData(TOP_OFFSET + 54, NaN, BOTTOM_OFFSET, EDGE_OFFSET);
        _button.layoutData      = new AnchorLayoutData(TOP_OFFSET, EDGE_OFFSET);

        addChild(_nameLabel);
        addChild(_descLabel);
        addChild(_button);
    }

    public function get button():GlowFrameTextButton { return _button; }

    public function get data():Object { return _data; }
    public function set data(value:Object):void {
        _data = OptionSettings(value);

        if(_data == null) {
            _nameLabel.text         = "";
            _descLabel.text         = "";
            _button.textField.text  = "";
        }
        else {
            _nameLabel.text         = _data.name;
            _descLabel.text         = _data.description;
            _button.textField.text  = _data.availableValues[_data.currentValueIndex];
        }
    }

    public function get index():int { return _index; }
    public function set index(value:int):void { _index = value; }

    public function get owner():List { return _owner; }
    public function set owner(value:List):void { _owner = value; }

    public function get factoryID():String { return null; }
    public function set factoryID(value:String):void { }

    public function get isSelected():Boolean { return false; }
    public function set isSelected(value:Boolean):void { }

    override public function render(support:RenderSupport, parentAlpha:Number):void {
        redrawBackground();

        support.pushMatrix();
        support.transformMatrix(_background);
        support.blendMode = _background.blendMode;

        _background.render(support, parentAlpha);

        support.blendMode = blendMode;
        support.popMatrix();

        super.render(support, parentAlpha);
    }

    override protected function draw():void {
        if(actualWidth != actualWidth || actualWidth <= 0)
            return;

        _descLabel.width = actualWidth - 2 * EDGE_OFFSET;
        _descLabel.height = NaN;

        super.draw();
    }

    private function redrawBackground():void {
        var w:Number = width;
        var h:Number = height;

        var xOffset:Number = 4.33, xMargin:Number = 4, yOffset:Number = 6;

        if(w == 0 || h == 0)
            return;

        if(_background != null && Math.abs(_background.width - w + 2 * xMargin) < 1 && Math.abs(_background.height - h) < 1)
            return;

        if(_background != null)
            _background.dispose();

        _background = new GlowPanel(
            new Scale3Textures(_assetManager.getTexture("ui/thick_window_edge"), 6, 2),
            new Point2D(xOffset, yOffset),
            _assetManager.getTexture("ui/window_gradient"),
            _assetManager.getTexture("ui/window_pattern"),
            getScaledPolygon(w - xOffset - 2 * xMargin, h - yOffset)
        );

        _background.x = xMargin + xOffset / 2;
        _background.y = yOffset / 2;
    }
}
```

It's a shame you no longer support v2.3 - I think many of people are still using it, because it's hard, sometimes impossible (like in my case) to switch to the most recent Starling/Feathers. Daniel's policy about making bug-fixes and changes to the v1.8 is completely different, I never got a single response from him like the one I got from you, which sadly adds to my already long list of Feathers disappointments.

I don't see how I would ever consider using it again in a commercial project - the development period is rarely below 2 years and during that time you'd discontinue at least two different versions (!) of Feathers, so if I started with, lets say, 4.0 in the middle on my project 4.1 would come out and I'd be left with no way of fixing a bug. And I'm not even mentioning maintenance and further development after the project has been completed.

**Comment by joshtynjala** [comment]
I'm sorry to hear that. Unfortunately, my time is limited, and I need to work on other things to earn enough to pay my bills. As things are, I'm quite busy working on both bugs and features for each new version. Well, anyway, best of luck in the future then!
